### ISSUE TITLE
azure.iot.device.common.transport_exceptions.ConnectionDroppedError: None caused by Paho returned rc==1 caused by None

### ISSUE THREAD

**Issue by openedhardware** [Azure/azure-iot-sdk-python#445]
SDK Version: 2.0.0 like this:
```
root@raspberrypi:/tmp# pip freeze
azure-iot-device==2.0.0
certifi==2019.11.28
chardet==3.0.4
edgetpu==2.12.2
idna==2.8
janus==0.4.0
netifaces==0.10.9
numpy==1.16.2
opencv-contrib-python==4.1.1.26
paho-mqtt==1.5.0
Pillow==5.4.1
redis==3.3.11
requests==2.22.0
requests-unixsocket==0.2.0
six==1.14.0
transitions==0.7.2
urllib3==1.25.8
```

I have just used the sample sender/receiver examples here:
https://github.com/Azure/azure-iot-sdk-python-preview/blob/master/azure-iot-device/samples/legacy-samples/send_message_to_output.py
https://github.com/Azure/azure-iot-sdk-python-preview/blob/master/azure-iot-device/samples/legacy-samples/receive_message_on_input.py

But my scripts output several error messages like these:
![image](https://user-images.githubusercontent.com/16890683/73144107-ce849680-40dc-11ea-8961-becf36313fcc.png)
![image](https://user-images.githubusercontent.com/16890683/73144112-d47a7780-40dc-11ea-97f6-931854abcde0.png)

Any idea?

By the way, I had set the `networking mode` of this module as `host` by following this - https://github.com/Azure/iot-edge-v1/issues/517#issuecomment-407453239

**Comment by olivakar** [comment]
For the sender sample , were you able to actually send messages but saw errors on terminal ?
Or were you unable to send messages due to the errors ?

**Comment by openedhardware** [comment]
@olivakar 
I couldn't see any error message from the sender function.

These errors are displayed randomly on sender & receiver...

**Comment by olivakar** [comment]
so again to be absolutely sure in both scenarios (Sender and Receiver) you were able to send messages and receive messages ?

**Comment by openedhardware** [comment]
Sender was executed without any error message except the random errors above.
The receiver was waiting for the incoming message, but didn't receive anything and displayed random errors above again.

**Comment by openedhardware** [comment]
@olivakar 
Please check this video file.
https://my.pcloud.com/publink/show?code=XZ0ww2kZ6cMBCQhTtlfsaN86jl6KTjtDAPJX

Note that `sender.py` & `receiver.py` were exected on the **same** module.

**Comment by BertKleewein** [comment]
@openedhardware - when you say you were executing on the same module, do you mean that both processes were running at the same time?

I'm also a little confused about your description because a comment above says that you set the networking mode on your docker container to "host" , but your screenshots above show you running the code from a bash prompt that doesn't appear to be inside a docker container.   Is this correct?

**Comment by openedhardware** [comment]
@BertKleewein 
Yes, both were executed on the **same module** for sure.

Nah, I had executed them from a bash on an IoT Edge **module**.
Just tried `docker ps` to get the ID of my module, and switched to that module by `docker exec -it <id> bash`

Cheers.

**Comment by BertKleewein** [comment]
@openedhardware - Thank you for the video.  That really helps.  The problem that you're seeing is caused by the way you're using the single module identity.  

An IoT Edge Module can only be used by a single process at a time.  When you call create_from_environment() or create_from_connection_string(), you're preparing a new network connection for your module.  When you call connect() or send_event() or any other method that neds to use the network, it establishes the connection.  If a different process already has a connection for the module, that connection gets closed by the server because there can only be one connection.

As a side-effect of this, if the processes are aggressive about keeping the connection alive (like this SDK is), there can be a ping-pong effect between the 2 processes.  Say the first process is connected, then the second process connects which causes the first process's connect ion to close.  The first process notices the error and tries to reconnect, which causes the second process to disconnect.  Then, the second process tries to reconnect which causes the first process to disconnect, and so on.  This may be part of what you're seeing.

The other factor is that the docker container for your module probably has a process that's has it's own network connection for the module.  Or maybe not, I can't say.  But, if it does, that means you have three different processes all fighting to be the one that's connected.

If you want to test sender and receiver together like this, you need two different modules and a routing rule to connect the output of one module to the input of the other module.  You also need to ensure that that your don't have any other processes trying to connect.  This means creating your own docker images, either a single image with bash as an entrypoint, or two images, one with send.py and the other with receive.py as entrypoints.

I hope this helps.  Please let me know if this isn't clear.

**Comment by openedhardware** [comment]
@BertKleewein 
Thanks for your detailed explanation!

Alright, I am going to follow your instructions now and will get back to you soon.

Cheers.

**Comment by openedhardware** [comment]
@BertKleewein 

Hmm, I still cannot send/receive messages between modules... 😕 

I had commented out my python script in the `Dockerfile.arm32v7` and added `sleep` function to test manually on bash:
```
CMD [ "sleep", "infinity"]
# CMD [ "python3", "-u", "./main.py" ]
```
Having a couple of modules running this `CMD [ "sleep", "infinity"]` command(deployed and pulled), I SSH'ed into these modules and tried to send/receive with the sample scripts above...
Of course, the random error messages were gone, but `sender.py` didn't receive anything..

Do I have to configure something in the module to send/receive messages between modules on a single device?

Cheers.

**Comment by BertKleewein** [comment]
Yes, you have to declare a route to send/receive between modules.  If you want to send from module1 to module2, you might use this rule to connect the module1 output named "toModule2" to the module2 input named "fromModule1".
```
FROM /messages/modules/module1/outputs/toModule2 INTO BrokeredEndpoint("/modules/module2/inputs/fromModule1")
```
You set the routes in the Azure portal by clicking the "Set Modules" button and going to the "Routes" page.
